src/KubernetesClient/Informers/ResourceObservableExtensions.cs

```diff
@@ -0,0 +1,294 @@
+                            return;
+
+                        foreach (var item in resetSelector(resetBuffer))
+                            observer.OnNext(itemSelector(item));
+                        resetBuffer.Clear();
+                    }
+
+                    void OnComplete()
+                    {
+                        FlushBuffer();
+                        observer.OnCompleted();
+                    }
+
+                    void OnError(Exception e)
+                    {
+                        observer.OnError(e);
+                    }
+
+                    var upstreamSubscription = source
+                        .Subscribe(notification =>
+                        {
+                            if (notification.EventFlags.HasFlag(EventTypeFlags.Reset))
+                            {
+                                resetBuffer.Add(notification);
+                                if (!notification.EventFlags.HasFlag(EventTypeFlags.ResetEnd)) // continue buffering till we reach the end of list window
+                                    return;
+                            }
+
+                            if (notification.EventFlags.HasFlag(EventTypeFlags.ResetEnd))
+                            {
+                                FlushBuffer();
+                                return;
+                            }
+
+                            if (!notification.EventFlags.HasFlag(EventTypeFlags.Reset) && resetBuffer.Count > 0) FlushBuffer();
+
+                            observer.OnNext(itemSelector(notification));
+                        }, OnError, OnComplete);
+                    return new CompositeDisposable(upstreamSubscription, Disposable.Create(OnComplete));
+                })
+                .ObserveOn(Scheduler.Immediate);
+        }
+
+        /// <summary>
+        ///     Synchronizes the specified cache with resource event stream such that cache is maintained up to date.
+        /// </summary>
+        /// <param name="source">The source sequence</param>
+        /// <param name="cache">The cache to synchronize</param>
+        /// <param name="keySelector">The key selector function</param>
+        /// <typeparam name="TKey">The type of key</typeparam>
+        /// <typeparam name="TResource">The type of resource</typeparam>
+        /// <returns>Source sequence wrapped into <see cref="CacheSynchronized{T}" />, which allows downstream consumers to synchronize themselves with cache version</returns>
+        public static IObservable<CacheSynchronized<ResourceEvent<TResource>>> SynchronizeCache<TKey, TResource>(
+            this IObservable<ResourceEvent<TResource>> source,
+            ICache<TKey, TResource> cache,
+            Func<TResource, TKey> keySelector,
+            Func<TResource, TResource, TResource> mapper = null)
+        {
+            if (mapper == null)
+                mapper = (oldResource, newResource) => newResource;
+            long msgNum = 0;
```

Comment:
`msgNum` will be shared across all subscriptions and I do not believe that is correct.

Consider using `Observable.Defer` to close over `msgNum` instead.